src/type_seed.jl

```diff
@@ -0,0 +1,105 @@
+"""
+    type_seed(x)
+
+Computes a value to use as a seed for computing the hash value of a type.
+
+The constants used in this computation are random numbers
+produced by `Random.rand(Random.RandomDevice(), UInt)`.
+"""
+function type_seed end
+
+function type_seed(m::Module, h::UInt)
+    if m === Main || m === Base || m === Core
+        return h
+    else
+        mp = parentmodule(m)
+        if mp === m
+            return h
+        else
+            return type_seed(nameof(m), type_seed(mp, h))
+        end
+    end
+end
+
+function type_seed(t::UnionAll, h::UInt)
+    h = hash(h, 0xbd5f3e4941dba79d)
+    h = type_seed(t.var, h)
+    h = type_seed(t.body, h)
+    return h
+end
+
+function type_seed(x::Union, h::UInt)
+    h = hash(h, 0xa16a31201c4852c2)
+    for t in Base.uniontypes(x)
+        h = type_seed(t, h)
+    end
+    return h
+end
+
+function type_seed(t::TypeVar, h::UInt)
+    h = hash(h, 0xc921c42a65aee273)
+    h = type_seed(t.name, h)
+    h = type_seed(t.lb, h)
+    h = type_seed(t.ub, h)
+    return h
+end
+
+function type_seed(t::DataType, h::UInt)
```

Comment:
This implementation seems conceptually good, but do you know how much costlier it is compared to `Base.hash(type)`? Although I only care because it seems like we run this every time we hash an instance of an `@auto_hash_equals` type, see my other comment above.

Reply:
No, I don't know.  But since `Base.hash` isn't stable, that isn't really an option, is it?

This isn't used by default.  It is only used if you ask the type to be included but don't provide your own type seed function.

Reply:
Ok I followed up on Slack with a clarification question regarding our internal use